File: bulk_reactions.py
#!/usr/bin/env python3
"""Send the same reaction to many messages in a chat."""

from __future__ import annotations

import argparse
import asyncio
import os
import re
from pathlib import Path
from typing import Iterable

from dotenv import load_dotenv
from telethon.tl.functions.messages import SendReactionRequest
from telethon.tl.types import ReactionEmoji

from chat_export_utils import load_chats_from_config
from telegram_session import SessionLock, get_client

load_dotenv()


def parse_message_token(token: str) -> int | None:
    """Extract message id from a number or t.me link."""
    cleaned = token.strip()
    if not cleaned:
        return None
    if cleaned.isdigit():
        return int(cleaned)

    match = re.search(r"/(\d+)(?:\?.*)?$", cleaned)
    if match:
        return int(match.group(1))
    return None


def collect_message_ids(raw_ids: Iterable[str], file_path: Path | None) -> list[int]:
    """Combine ids from CLI and file, allowing comma/space-separated tokens."""
    tokens: list[str] = []
    for raw in raw_ids:
        tokens.extend(part for part in re.split(r"[,\s]+", raw.strip()) if part)

    if file_path:
        for line in file_path.read_text(encoding="utf-8").splitlines():
            tokens.extend(part for part in re.split(r"[,\s]+", line.strip()) if part)

    ids: list[int] = []
    for token in tokens:
        msg_id = parse_message_token(token)
        if msg_id is not None:
            ids.append(msg_id)
    return ids


def resolve_chat_target(chat_arg: str | None):
    """Resolve chat alias/id from CLI/env into something Telethon can use."""
    chat = chat_arg or os.getenv("BULK_REACTIONS_CHAT") or os.getenv("LIKE_CHAT_ID")
    if not chat or not str(chat).strip():
        raise SystemExit("Set --chat or BULK_REACTIONS_CHAT/LIKE_CHAT_ID env var.")

    chat = str(chat).strip()

    aliases = load_chats_from_config("config.yml", silent_missing=True, skip_invalid=True)
    if chat in aliases:
        return aliases[chat]
    if chat.startswith("@"):
        return chat
    if chat.lstrip("-").isdigit():
        return int(chat)
    return chat


async def fetch_recent_ids(client, entity, limit: int, user_id: int | None) -> list[int]:
    """Collect last N message ids, optionally filtering by sender."""
    messages = await client.get_messages(entity, limit=limit)
    return [m.id for m in messages if not user_id or m.sender_id == user_id]


async def build_previews(client, entity, ids: list[int]) -> dict[int, str]:
    """Fetch messages once to show short previews alongside ids."""
    previews: dict[int, str] = {}
    fetched = await client.get_messages(entity, ids=ids)
    for msg in fetched:
        if not msg:
            continue
        text = (msg.text or msg.raw_text or "").replace("\n", " ").strip()
        label = text[:80] + ("…" if len(text) > 80 else "")
        previews[msg.id] = label or "[no text]"
    return previews


async def main():
    parser = argparse.ArgumentParser(description="Send reactions to many messages.")
    parser.add_argument(
        "--chat",
        help="Chat alias from config.yml, @username or numeric id (fallback: BULK_REACTIONS_CHAT or LIKE_CHAT_ID env).",
    )
    parser.add_argument(
        "--emoji",
        default=os.getenv("BULK_REACTIONS_EMOJI", "🔥"),
        help="Reaction emoji to send (default: 🔥 or BULK_REACTIONS_EMOJI env).",
    )
    parser.add_argument(
        "--ids",
        nargs="*",
        default=[],
        help="Message ids or t.me links (space/comma separated).",
    )
    parser.add_argument(
        "--file",
        type=Path,
        help="Path to file with message ids or t.me links (one per line).",
    )
    parser.add_argument(
        "--recent",
        type=int,
        default=0,
        help="Also react to last N messages from chat (optional).",
    )
    parser.add_argument(
        "--user-id",
        type=int,
        help="When used with --recent, only react to messages from this sender id.",
    )
    parser.add_argument(
        "--delay",
        type=float,
        default=float(os.getenv("BULK_REACTIONS_DELAY", "0.6")),
        help="Delay in seconds between reactions to avoid rate limits (default: 0.6s).",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Preview actions without sending reactions.",
    )

    args = parser.parse_args()

    chat_target = resolve_chat_target(args.chat)
    message_ids = collect_message_ids(args.ids, args.file)
    if not message_ids and args.recent <= 0:
        raise SystemExit("Provide message ids/links or set --recent to react to latest messages.")

    with SessionLock():
        client = get_client()
        async with client:
            entity = await client.get_entity(chat_target)

            if args.recent > 0:
                recent_ids = await fetch_recent_ids(client, entity, args.recent, args.user_id)
                message_ids.extend(recent_ids)

            unique_ids: list[int] = []
            seen = set()
            for msg_id in message_ids:
                if msg_id not in seen:
                    seen.add(msg_id)
                    unique_ids.append(msg_id)

            if not unique_ids:
                print("Nothing to do: no valid message ids resolved.")
                return

            previews = await build_previews(client, entity, unique_ids)

            print(f"Chat: {chat_target}")
            print(f"Emoji: {args.emoji}")
            print(f"Messages to react: {len(unique_ids)}")
            if args.dry_run:
                print("Dry run: no reactions will be sent.")
            print()

            sent = 0
            errors = 0
            for msg_id in unique_ids:
                preview = previews.get(msg_id, "[message not found]")
                print(f"{args.emoji} -> {msg_id}: {preview}")

                if args.dry_run:
                    continue

                try:
                    await client(
                        SendReactionRequest(
                            peer=entity,
                            msg_id=msg_id,
                            reaction=[ReactionEmoji(emoticon=args.emoji)],
                            big=False,
                            add_to_recent=False,
                        )
                    )
                    sent += 1
                    await asyncio.sleep(args.delay)
                except Exception as exc:
                    errors += 1
                    print(f"  Error on {msg_id}: {exc}")

            print()
            print("Done.")
            print(f"Sent: {sent}")
            print(f"Errors: {errors}")


if __name__ == "__main__":
    asyncio.run(main())

File: src/bin/bulk_reactions.rs
//! Send the same reaction to many messages in a chat.
//!
//! Usage:
//!   cargo run --bin bulk_reactions -- --chat @channel --emoji 🔥 --ids 123 456 789
//!   cargo run --bin bulk_reactions -- --chat @channel --recent 50

use anyhow::Result;
use clap::Parser;
use dotenvy::dotenv;
use grammers_client::types::peer::Peer;
use grammers_tl_types as tl;
use regex::Regex;
use std::collections::HashSet;
use std::env;
use std::fs;
use std::path::PathBuf;
use std::time::Duration;
use telegram_reader::chat::find_chat;
use telegram_reader::get_client;
use telegram_reader::session::SessionLock;
use tokio::time::sleep;
use tracing::error;

/// Convert a Peer to InputPeer for API calls.
fn peer_to_input(peer: &Peer) -> tl::enums::InputPeer {
    match peer {
        Peer::User(user) => {
            let (user_id, access_hash) = match &user.raw {
                tl::enums::User::User(u) => (u.id, u.access_hash.unwrap_or(0)),
                tl::enums::User::Empty(u) => (u.id, 0),
            };
            tl::enums::InputPeer::User(tl::types::InputPeerUser {
                user_id,
                access_hash,
            })
        }
        Peer::Channel(channel) => tl::enums::InputPeer::Channel(tl::types::InputPeerChannel {
            channel_id: channel.raw.id,
            access_hash: channel.raw.access_hash.unwrap_or(0),
        }),
        Peer::Group(group) => match &group.raw {
            tl::enums::Chat::Chat(c) => {
                tl::enums::InputPeer::Chat(tl::types::InputPeerChat { chat_id: c.id })
            }
            tl::enums::Chat::Channel(c) => {
                tl::enums::InputPeer::Channel(tl::types::InputPeerChannel {
                    channel_id: c.id,
                    access_hash: c.access_hash.unwrap_or(0),
                })
            }
            _ => tl::enums::InputPeer::Empty,
        },
    }
}

#[derive(Parser, Debug)]
#[command(name = "bulk_reactions")]
#[command(about = "Send reactions to many messages in a chat")]
struct Args {
    /// Chat alias from config.yml, @username or numeric id
    #[arg(long, env = "BULK_REACTIONS_CHAT")]
    chat: Option<String>,

    /// Reaction emoji to send
    #[arg(long, env = "BULK_REACTIONS_EMOJI", default_value = "🔥")]
    emoji: String,

    /// Message ids or t.me links (space separated)
    #[arg(long, num_args = 0..)]
    ids: Vec<String>,

    /// Path to file with message ids or t.me links
    #[arg(long)]
    file: Option<PathBuf>,

    /// React to last N messages from chat
    #[arg(long, default_value = "0")]
    recent: usize,

    /// Only react to messages from this sender id (with --recent)
    #[arg(long)]
    user_id: Option<i64>,

    /// Delay in seconds between reactions
    #[arg(long, env = "BULK_REACTIONS_DELAY", default_value = "0.6")]
    delay: f64,

    /// Preview actions without sending reactions
    #[arg(long)]
    dry_run: bool,
}

/// Parse message id from a number or t.me link.
fn parse_message_token(token: &str) -> Option<i32> {
    let cleaned = token.trim();
    if cleaned.is_empty() {
        return None;
    }

    // Try parsing as number
    if let Ok(id) = cleaned.parse::<i32>() {
        return Some(id);
    }

    // Try extracting from t.me link
    let re = Regex::new(r"/(\d+)(?:\?.*)?$").ok()?;
    re.captures(cleaned)
        .and_then(|c| c.get(1))
        .and_then(|m| m.as_str().parse().ok())
}

/// Collect message ids from CLI args and file.
fn collect_message_ids(raw_ids: &[String], file_path: Option<&PathBuf>) -> Vec<i32> {
    let mut tokens: Vec<String> = Vec::new();

    // Parse CLI args (comma/space separated)
    for raw in raw_ids {
        for part in raw.split(|c: char| c == ',' || c.is_whitespace()) {
            let part = part.trim();
            if !part.is_empty() {
                tokens.push(part.to_string());
            }
        }
    }

    // Parse file if provided
    if let Some(path) = file_path {
        if let Ok(content) = fs::read_to_string(path) {
            for line in content.lines() {
                for part in line.split(|c: char| c == ',' || c.is_whitespace()) {
                    let part = part.trim();
                    if !part.is_empty() {
                        tokens.push(part.to_string());
                    }
                }
            }
        }
    }

    // Convert to ids
    tokens
        .iter()
        .filter_map(|t| parse_message_token(t))
        .collect()
}

/// Resolve chat target from args/env.
fn resolve_chat_target(chat_arg: Option<&str>) -> Result<String> {
    let chat = chat_arg
        .map(|s| s.to_string())
        .or_else(|| env::var("BULK_REACTIONS_CHAT").ok())
        .or_else(|| env::var("LIKE_CHAT_ID").ok());

    let chat = chat
        .ok_or_else(|| anyhow::anyhow!("Set --chat or BULK_REACTIONS_CHAT/LIKE_CHAT_ID env var"))?;

    Ok(chat.trim().to_string())
}

#[tokio::main]
async fn main() -> Result<()> {
    dotenv().ok();
    tracing_subscriber::fmt::init();

    let args = Args::parse();

    let chat_target = resolve_chat_target(args.chat.as_deref())?;
    let mut message_ids = collect_message_ids(&args.ids, args.file.as_ref());

    if message_ids.is_empty() && args.recent == 0 {
        anyhow::bail!("Provide message ids/links or set --recent to react to latest messages");
    }

    let _lock = SessionLock::acquire()?;
    let client = get_client().await?;

    // Resolve chat using find_chat (handles config aliases, usernames, numeric ids)
    let peer = find_chat(&client, &chat_target).await?;
    let chat_name = peer.name().unwrap_or("Unknown");

    // Fetch recent messages if requested
    if args.recent > 0 {
        let mut messages_iter = client.iter_messages(&peer);
        let mut count = 0;

        while let Some(message) = messages_iter.next().await? {
            if count >= args.recent {
                break;
            }

            let sender_id = message.sender().map(|s| match s {
                grammers_client::types::peer::Peer::User(u) => u.raw.id(),
                grammers_client::types::peer::Peer::Channel(c) => c.raw.id,
                grammers_client::types::peer::Peer::Group(g) => {
                    // Extract id from group's raw enum
                    match &g.raw {
                        grammers_tl_types::enums::Chat::Empty(c) => c.id,
                        grammers_tl_types::enums::Chat::Chat(c) => c.id,
                        grammers_tl_types::enums::Chat::Forbidden(c) => c.id,
                        grammers_tl_types::enums::Chat::Channel(c) => c.id,
                        grammers_tl_types::enums::Chat::ChannelForbidden(c) => c.id,
                    }
                }
            });

            // Filter by user if specified
            if let Some(target_user) = args.user_id {
                if sender_id != Some(target_user) {
                    continue;
                }
            }

            message_ids.push(message.id());
            count += 1;
        }
    }

    // Deduplicate
    let mut seen = HashSet::new();
    let unique_ids: Vec<i32> = message_ids
        .into_iter()
        .filter(|id| seen.insert(*id))
        .collect();

    if unique_ids.is_empty() {
        println!("Nothing to do: no valid message ids resolved.");
        return Ok(());
    }

    // Build previews
    let mut previews: std::collections::HashMap<i32, String> = std::collections::HashMap::new();
    for &msg_id in &unique_ids {
        // Fetch message for preview
        let mut iter = client.iter_messages(&peer);
        while let Some(message) = iter.next().await? {
            if message.id() == msg_id {
                let text = message.text();
                let preview = if text.is_empty() {
                    "[no text]".to_string()
                } else if text.len() > 80 {
                    format!("{}…", &text[..80].replace('\n', " "))
                } else {
                    text.replace('\n', " ")
                };
                previews.insert(msg_id, preview);
                break;
            }
        }
    }

    println!("Chat: {}", chat_name);
    println!("Emoji: {}", args.emoji);
    println!("Messages to react: {}", unique_ids.len());
    if args.dry_run {
        println!("Dry run: no reactions will be sent.");
    }
    println!();

    let mut sent = 0;
    let mut errors = 0;

    for msg_id in unique_ids {
        let preview = previews
            .get(&msg_id)
            .map(|s| s.as_str())
            .unwrap_or("[message not found]");
        println!("{} -> {}: {}", args.emoji, msg_id, preview);

        if args.dry_run {
            continue;
        }

        // Send reaction using raw API
        let reaction = tl::enums::Reaction::Emoji(tl::types::ReactionEmoji {
            emoticon: args.emoji.clone(),
        });

        let request = tl::functions::messages::SendReaction {
            peer: peer_to_input(&peer),
            msg_id,
            reaction: Some(vec![reaction]),
            big: false,
            add_to_recent: false,
        };

        match client.invoke(&request).await {
            Ok(_) => {
                sent += 1;
                sleep(Duration::from_secs_f64(args.delay)).await;
            }
            Err(e) => {
                errors += 1;
                error!(msg_id = msg_id, error = %e, "Error sending reaction");
            }
        }
    }

    println!();
    println!("Done.");
    println!("Sent: {}", sent);
    println!("Errors: {}", errors);

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_message_token() {
        assert_eq!(parse_message_token("123"), Some(123));
        assert_eq!(parse_message_token("https://t.me/channel/456"), Some(456));
        assert_eq!(parse_message_token("t.me/c/123456/789"), Some(789));
        assert_eq!(parse_message_token(""), None);
        assert_eq!(parse_message_token("abc"), None);
    }
}
